### ISSUE TITLE
SQLAlchemy doesn't upgrade when python constraint > 3.9

### ISSUE THREAD

**Issue by wwuck** [renovatebot/renovate#19500]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

main branch `aab79f15d`

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the bug

Reproduction repository: https://github.com/wwuck/renovate-sqlalchemy-test

SQLAlchemy package on pypi is not updated by renovate when the renovate configuration has a python constraints value above 3.9.

In the reproduction repository, I have set
```json
"constraints": {
    "python": "3.10"
}
```
and sqlalchemy does not get any updates.

If I change to
```json
"constraints": {
    "python": "3.9"
}
```
then sqlalchemy is updated correctly by renovate.

The reproduction repository also includes the `Flake8-pyproject` package to check the problem is not with uppercase letters in the package name.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: packageFiles with updates (repository=wwuck/renovate-sqlalchemy-test, baseBranch=main)
       "config": {
         "pip_requirements": [
           {
             "packageFile": "requirements.txt",
             "deps": [
               {
                 "depName": "sqlalchemy",
                 "currentValue": "==1.4.32",
                 "datasource": "pypi",
                 "currentVersion": "1.4.32",
                 "depIndex": 0,
                 "updates": [],
                 "warnings": [],
                 "versioning": "pep440",
                 "registryUrl": "https://pypi.org/pypi",
                 "homepage": "https://www.sqlalchemy.org",
                 "changelogUrl": "https://docs.sqlalchemy.org/en/latest/changelog/",
                 "fixedVersion": "1.4.32"
               },
               {
                 "depName": "flake8-pyproject",
                 "currentValue": "==1.2.1",
                 "datasource": "pypi",
                 "currentVersion": "1.2.1",
                 "depIndex": 1,
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "1.2.2",
                     "newValue": "==1.2.2",
                     "releaseTimestamp": "2022-12-04T12:03:21.000Z",
                     "newMajor": 1,
                     "newMinor": 2,
                     "updateType": "patch",
                     "isRange": true,
                     "branchName": "renovate/flake8-pyproject-1.x"
                   }
                 ],
                 "warnings": [],
                 "versioning": "pep440",
                 "sourceUrl": "https://github.com/john-hen/Flake8-pyproject",
                 "registryUrl": "https://pypi.org/pypi",
                 "isSingleVersion": true,
                 "fixedVersion": "1.2.1"
               }
             ]
           }
         ]
       }
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

**Comment by rarkins** [comment]
Does sqlalchemy support Python 3.10?

**Comment by wwuck** [comment]
Yes. On the pypi page https://pypi.org/project/SQLAlchemy/ it shows support for 3.10 and 3.11.

**Comment by rarkins** [comment]
What's the difference in log messages between 3.9 (working) and 3.10 (not working)?

**Comment by wwuck** [comment]
`"constraints": { "python" : "3.9" }`:
```
DEBUG: packageFiles with updates (repository=wwuck/renovate-sqlalchemy-test, baseBranch=main)
       "config": {
         "pip_requirements": [
           {
             "packageFile": "requirements.txt",
             "deps": [
               {
                 "depName": "sqlalchemy",
                 "currentValue": "==1.4.32",
                 "datasource": "pypi",
                 "currentVersion": "1.4.32",
                 "depIndex": 0,
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "1.4.45",
                     "newValue": "==1.4.45",
                     "releaseTimestamp": "2022-12-10T18:57:26.000Z",
                     "newMajor": 1,
                     "newMinor": 4,
                     "updateType": "patch",
                     "isRange": true,
                     "branchName": "renovate/sqlalchemy-1.x"
                   }
                 ],
                 "warnings": [],
                 "versioning": "pep440",
                 "registryUrl": "https://pypi.org/pypi",
                 "homepage": "https://www.sqlalchemy.org",
                 "changelogUrl": "https://docs.sqlalchemy.org/en/latest/changelog/",
                 "isSingleVersion": true,
                 "fixedVersion": "1.4.32"
               },
               {
                 "depName": "flake8-pyproject",
                 "currentValue": "==1.2.1",
                 "datasource": "pypi",
                 "currentVersion": "1.2.1",
                 "depIndex": 1,
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "1.2.2",
                     "newValue": "==1.2.2",
                     "releaseTimestamp": "2022-12-04T12:03:21.000Z",
                     "newMajor": 1,
                     "newMinor": 2,
                     "updateType": "patch",
                     "isRange": true,
                     "branchName": "renovate/flake8-pyproject-1.x"
                   }
                 ],
                 "warnings": [],
                 "versioning": "pep440",
                 "sourceUrl": "https://github.com/john-hen/Flake8-pyproject",
                 "registryUrl": "https://pypi.org/pypi",
                 "isSingleVersion": true,
                 "fixedVersion": "1.2.1"
               }
             ]
           }
         ]
       }
```
`"constraints": { "python" : "3.10" }`:
```
DEBUG: packageFiles with updates (repository=wwuck/renovate-sqlalchemy-test, baseBranch=main)
       "config": {
         "pip_requirements": [
           {
             "packageFile": "requirements.txt",
             "deps": [
               {
                 "depName": "sqlalchemy",
                 "currentValue": "==1.4.32",
                 "datasource": "pypi",
                 "currentVersion": "1.4.32",
                 "depIndex": 0,
                 "updates": [],
                 "warnings": [],
                 "versioning": "pep440",
                 "registryUrl": "https://pypi.org/pypi",
                 "homepage": "https://www.sqlalchemy.org",
                 "changelogUrl": "https://docs.sqlalchemy.org/en/latest/changelog/",
                 "fixedVersion": "1.4.32"
               },
               {
                 "depName": "flake8-pyproject",
                 "currentValue": "==1.2.1",
                 "datasource": "pypi",
                 "currentVersion": "1.2.1",
                 "depIndex": 1,
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "1.2.2",
                     "newValue": "==1.2.2",
                     "releaseTimestamp": "2022-12-04T12:03:21.000Z",
                     "newMajor": 1,
                     "newMinor": 2,
                     "updateType": "patch",
                     "isRange": true,
                     "branchName": "renovate/flake8-pyproject-1.x"
                   }
                 ],
                 "warnings": [],
                 "versioning": "pep440",
                 "sourceUrl": "https://github.com/john-hen/Flake8-pyproject",
                 "registryUrl": "https://pypi.org/pypi",
                 "isSingleVersion": true,
                 "fixedVersion": "1.2.1"
               }
             ]
           }
         ]
       }
```
Is that sufficient? I can copy both log files in full if you need more.

**Comment by rarkins** [comment]
I debugged this and made these observations:

1. The versions are getting filtered out have constraints like `!=3.0.*,!=3.1.*,!=3.2.*,!=3.3.*,!=3.4.*,!=3.5.*,>=2.7`
2. `version.isVersion()` returns `true` for both 3.9 and 3.10
3. 3.9 matches the range, 3.10 does not

I added these tests locally for pep440 versioning to confirm what I was seeing:

```
  test.each`
    range                                                      | version     | expected
    ${'!=3.0.*,!=3.1.*,!=3.2.*,!=3.3.*,!=3.4.*,!=3.5.*,>=2.7'} | ${'3.9'}    | ${true}
    ${'!=3.0.*,!=3.1.*,!=3.2.*,!=3.3.*,!=3.4.*,!=3.5.*,>=2.7'} | ${'3.9.0'}  | ${true}
    ${'!=3.0.*,!=3.1.*,!=3.2.*,!=3.3.*,!=3.4.*,!=3.5.*,>=2.7'} | ${'3.10'}   | ${false}
    ${'!=3.0.*,!=3.1.*,!=3.2.*,!=3.3.*,!=3.4.*,!=3.5.*,>=2.7'} | ${'3.10.0'} | ${false}
    ${'!=3.0.*,!=3.2.*,!=3.3.*,!=3.4.*,!=3.5.*,>=2.7'}         | ${'3.10'}   | ${true}
  `(
    'matches("$range", "$version") === $expected',
    ({ range, version, expected }) => {
      expect(pep440.matches(version, range)).toBe(expected);
    }
  );
```

I would have expected them all to be `true` though.

So it seems like it could be a bug in our pep440 versioning.

What is your reason for setting `constraints.python`? e.g. is it so that you restrict upgrades to those supporting 3.10?

**Comment by wwuck** [comment]
Yes, we run on a specific python version and need to validate all packages are supported before we can upgrade our environment to the next python major version. We are currently moving to 3.11 and only added the constraints config recently. We didn’t have the constraints config set previously when we were on 3.9.

**Comment by rarkins** [comment]
Fixed by #19528
